### ISSUE TITLE
[$500] Do not allow emojis to be formatted (ie italics)

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14676]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to android app
2. Go to any chat
3. Try to make emoji in italic form:-Type in :  _insertemoji_  For example:_⁦:100:_,  _⁦:purple_heart:_  , _⁦:partying_face:_ ,
4. Notice that the top right part of the emoji cuts off. Few Example: the :100:, :purple_heart:, :partying_face: emoji has its top right section cut , (It happens on android only latest)

## Expected Result:
Users should not be able to format emojis, lets keep them with normal style (no bold, italics, strikethrough etc)

## Actual Result:
Users can make emoji italic which leads to the emojis to cut off a few parts of the emoji (for some emojis)

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] MacOS Desktop
- [x] MacOS Chrome Web

**Version Number:** 1.2.62-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

<img src="https://user-images.githubusercontent.com/43996225/215615747-0dfbebc7-7125-47ec-b737-6a0e909a2b65.jpg" width="40%" height="40%">

<img src="https://user-images.githubusercontent.com/43996225/215615757-791cca99-bb77-48d5-8adf-ebcbec38983b.jpg" width="40%" height="40%">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675094271396099

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017d08e0c09284d31f</li>
        <li>Upwork Job ID: 1620563000799895552</li>
        <li>Last Price Increase: 2024-08-08</li>
<li>Automatic offers: </li>
<ul>
<li>eh2077 | Reviewer | 0</li>
<li>wildan-m | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @mallenexpensify / @mallenexpensify</details>

**Comment by anmurali** [comment]
Confirmed that this happens on Android native app. 

Confirmed it does not happen on Mac/Chrome. But related, when I try the same reproduction steps on iOS native, iOS Chrome and Mac Safari - I cannot italicize emojis at all. Is that expected?

**Comment by alexxxwork** [comment]
###Proposal
Quick fix:
```diff
diff --git a/src/styles/styles.js b/src/styles/styles.js
index b826ed336f..84286591e9 100644
--- a/src/styles/styles.js
+++ b/src/styles/styles.js
@@ -63,6 +63,7 @@ const webViewStyles = {
         em: {
             fontFamily: fontFamily.EXP_NEUE,
             fontStyle: 'italic',
+            width:23
         },
 
         del: {
```

**Comment by flodnv** [comment]
I don't think we should be able to italicize emojis, asked here: https://expensify.slack.com/archives/C01GTK53T8Q/p1675266083150699

**Comment by JmillsExpensify** [comment]
Agree with @flodnv.

**Comment by tienifr** [comment]
# Proposal

The Emoji is italicized because it is contained in the `<em></em>` tag (same as when we make a text italic). We can observe the behavior on both Chrome web and Android.

In order to remove the ability to italicize emojis, I'll wrap any emoji that are inside the `<em></em>` with a `expensify-emoji` custom HTML tag, and give it a `fontStyle: 'normal'` style. This is the same way that Github does it.

(The emoji regex below is taken from the `CONST.REGEX.EMOJIS` in Expensify app)

In https://github.com/Expensify/expensify-common
```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index 7f94ffe..057e87c 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -134,6 +134,11 @@ export default class ExpensiMark {
                 regex: /(?!_blank")\b_((?!\s).*?\S)_\b(?![^<]*(<\/pre>|<\/code>|<\/a>|_blank))/g,
                 replacement: '<em>$1</em>',
             },
+            {
+                name: 'emoji',
+                regex: /(<em>.*?)([\p{Extended_Pictographic}\u200d\u{1f1e6}-\u{1f1ff}\u{1f3fb}-\u{1f3ff}\u{e0020}-\u{e007f}\u20E3\uFE0F]|[#*0-9]\uFE0F?\u20E3)(.*?<\/em>)/gu,
+                replacement: '$1<expensify-emoji>$2</expensify-emoji>$3',
+            },
             {
                 // Use \B in this case because \b doesn't match * or ~.
                 // \B will match everything that \b doesn't, so it works
                 
```

In https://github.com/Expensify/App

```diff
diff --git a/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js b/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
index 456cd1ecf3..0bf75ea9d2 100755
--- a/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
+++ b/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
@@ -39,6 +39,10 @@ const customHTMLElementModels = {
         tagName: 'email-comment',
         mixedUAStyles: {whiteSpace: 'normal'},
     }),
+    'expensify-emoji': defaultHTMLElementModels.span.extend({
+        tagName: 'expensify-emoji',
+        mixedUAStyles: {fontStyle: 'normal'},
+    }),
 };
 
 const defaultViewProps = {style: [styles.alignItemsStart, styles.userSelectText]};

```

There're small possible variations based on our preference:
- We don't have to use a custom `expensify-emoji` tag but just use a `span` tag with a style. I prefer using a custom tag to make it clear

There's a minor case when only there's only the emoji inside the `em`, no other text, the emoji will not be italicized but it will also not have the `onlyEmojisText` style, we can fix that with another regex.

Pls note we have to fix the same in the BE as well, in order to test the above fix we need to enable offline mode.

**Comment by tienifr** [comment]
working well after the fix:

https://user-images.githubusercontent.com/113963320/216310065-25e96b0c-aa25-4f56-bd73-79cd9d89c878.mov

**Comment by s77rt** [comment]
@tienifr Your approach looks solid however the regex is not correct as it will only match one emoji, try sending `hi 😁 there 😁`, it will wrap the first emoji but not the second one. Also it will only work with `em` tags and I think it would be better to wrap the emojis all the time (this will also cover the bold case).

Here are my suggestions:
1. Replace the regex part with:
```js
            {
                name: 'emoji',
                regex: /[\p{Extended_Pictographic}\u200d\u{1f1e6}-\u{1f1ff}\u{1f3fb}-\u{1f3ff}\u{e0020}-\u{e007f}\u20E3\uFE0F]|[#*0-9]\uFE0F?\u20E3/gu,
                replacement: match => `<expensify-emoji>${match}</expensify-emoji>`,
            },

```
2. Move the regex `emoji` to be before `italic` and after `autolink` (Not necessary, just a personal opinion as I like to preserve the order of italic, bold, strikethrough regex rules)
3. In `mixedUAStyles: {fontStyle: 'normal'}` add `fontWeight: 'normal'`

**Comment by tienifr** [comment]
Thanks @s77rt for your suggestion! Here's the my updated diff for the regex so that it matches all rather than just one.

```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index c535048..62b183f 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -242,6 +242,11 @@ export default class ExpensiMark {
                 regex: /<(em|i)(?:"[^"]*"|'[^']*'|[^'">])*>(.*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,
                 replacement: '_$2_',
             },
+            {
+                name: 'emoji',
+                regex: /([\p{Extended_Pictographic}\u200d\u{1f1e6}-\u{1f1ff}\u{1f3fb}-\u{1f3ff}\u{e0020}-\u{e007f}\u20E3\uFE0F]|[#*0-9]\uFE0F?\u20E3)/gu,
+                replacement: '<expensify-emoji>$1</expensify-emoji>',
+            },
             {
                 name: 'bold',
                 regex: /<(b|strong)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,

```

Regarding the `fontWeight: 'normal'`, I don't feel strongly about since it doesn't seem that the bold fontWeight affects how the emoji looks.

**Comment by tienifr** [comment]
@s77rt  Regarding the `fontWeight: 'normal'`, I don't feel strongly about since it doesn't seem that the bold fontWeight affects how the emoji looks.

**Comment by s77rt** [comment]
@tienifr 
- You can use the same regex as provided, there is no need to create a capturing group that captures the whole match (unnecessary doubling memory usage)
- If you bold the emoji it will be bold but it's really hard to notice the difference, the outer circle shadow will appear a bit thicker. Adding `fontWeight: 'normal'` ensures we are rendering the emoji correctly

**Comment by tienifr** [comment]
@s77rt I tried to zoom the emojis in but can't recognize any difference between an emoji that has `fontWeight: 'bold'` and a normal one. But we surely can add that if we want to be extra safe in that bold case (Github also does it).

Thanks for the input!

```diff
diff --git a/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js b/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
index 456cd1ecf3..0bf75ea9d2 100755
--- a/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
+++ b/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
@@ -39,6 +39,10 @@ const customHTMLElementModels = {
         tagName: 'email-comment',
         mixedUAStyles: {whiteSpace: 'normal'},
     }),
+    'expensify-emoji': defaultHTMLElementModels.span.extend({
+        tagName: 'expensify-emoji',
+        mixedUAStyles: {fontStyle: 'normal', fontWeight: 'normal'},
+    }),
 };
 
 const defaultViewProps = {style: [styles.alignItemsStart, styles.userSelectText]};
```

**Comment by s77rt** [comment]
@tienifr While seaching I found this image.
<img width="764" alt="yLrBRQd" src="https://user-images.githubusercontent.com/16493223/216660138-f60b1b1b-8a5a-4fda-8865-fcb0f6bb4203.png">

The middle emoji is in bold but it's hard to notice also it may depend on the font. Adding `fontWeight: 'normal'` is more a safety step.

**Comment by MelvinBot** [comment]
@flodnv, @anmurali, @aimane-chnaif Eep! 4 days overdue now. Issues have feelings too...

**Comment by MelvinBot** [comment]
@flodnv, @anmurali, @aimane-chnaif Huh... This is 4 days overdue. Who can take care of this?

**Comment by bernhardoj** [comment]
### **Proposal**
I was trying to find a solution for [this](https://github.com/Expensify/App/issues/4114) and stumbled upon this issue. I would like to propose an alternative solution that would be useful later.

I agree with the above proposal to enclose all emoji with a custom tag to apply a custom style later on.

But the first thing to do is to change the emoji regex.

I previously worked on this [issue](https://github.com/Expensify/App/issues/13823) by providing a new regex and it works well for that case. However, when we use that regex to match composite emojis, for example 👨‍👨‍👦, it will produce this array of string:
`['👨', '‍', '👨', '‍', '👦']`

The empty string is zero-width joiner. Again, this works well for the case I mentioned from the previous paragraph (we only care what the string contains). But for this case, we need to replace the emoji and enclose it with a tag. If we use the current emoji regex, it will enclose each part of the emoji with the tag, instead of the composited emoji.

Expected:
`<emoji>👨‍👨‍👦</emoji>`

Actual:
`<emoji>👨</emoji><emoji></emoji><emoji>👨</emoji> and so on...`

So, we need a better regex to capture the composite emoji, below is the new regex.

```diff
diff --git a/src/CONST.js b/src/CONST.js
index ae3ffca7f..8399de047 100755
--- a/src/CONST.js
+++ b/src/CONST.js
@@ -790,7 +790,7 @@ const CONST = {
         WEBSITE: /^((https?|ftp):\/\/)(([a-z\d]([a-z\d-]*[a-z\d])*)\.)+[a-z]{2,}(:\d+)?(\/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(#[-a-z\d_]*)?$/i,
 
         // eslint-disable-next-line max-len, no-misleading-character-class
-        EMOJIS: /[\p{Extended_Pictographic}\u200d\u{1f1e6}-\u{1f1ff}\u{1f3fb}-\u{1f3ff}\u{e0020}-\u{e007f}\u20E3\uFE0F]|[#*0-9]\uFE0F?\u20E3/gu,
+        EMOJIS: /[\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3/gu,
         TAX_ID: /^\d{9}$/,
         NON_NUMERIC: /\D/g,
         EMOJI_NAME: /:[\w+-]+:/g,
```

I have tested the new regex against all emojis from `assets/emojis` file and make sure all `EmojiTest.js` test cases pass.

Here is how I test it:

```diff
diff --git a/tests/unit/EmojiTest.js b/tests/unit/EmojiTest.js
index 25626872c..b68bfbca0 100644
--- a/tests/unit/EmojiTest.js
+++ b/tests/unit/EmojiTest.js
@@ -1,5 +1,6 @@
 import _ from 'underscore';
 import Emoji from '../../assets/emojis';
+import CONST from '../../src/CONST';
 import * as EmojiUtils from '../../src/libs/EmojiUtils';
 
 describe('EmojiTest', () => {
@@ -11,12 +12,16 @@ describe('EmojiTest', () => {
             }
 
             // When we match every Emoji Code
-            const isEmojiMatched = EmojiUtils.containsOnlyEmojis(emoji.code);
+            const isEmojiMatched = EmojiUtils.containsOnlyEmojis(emoji.code) 
+                                && emoji.code.match(CONST.REGEX.EMOJIS).length === 1;
 
             let skinToneMatched = true;
             if (emoji.types) {
                 // and every skin tone variant of the Emoji code
-                skinToneMatched = _.every(emoji.types, emojiWithSkinTone => EmojiUtils.containsOnlyEmojis(emojiWithSkinTone));
+                skinToneMatched = _.every(emoji.types, emojiWithSkinTone => {
+                    return EmojiUtils.containsOnlyEmojis(emojiWithSkinTone) 
+                        && emoji.code.match(CONST.REGEX.EMOJIS).length === 1;
+                });
             }
             return skinToneMatched && isEmojiMatched;
         });

```

Here is the code part to enclose the emoji with <emoji> tag. 

<details>

<summary>explanation</summary>

I prefer to do the string replacement here compared to doing it on `ExpensiMark`. As I said at the first sentence above, this will make easier when we do the replacement inside our App code for [this](https://github.com/Expensify/App/issues/4114) improvement. For the context, the mentioned improvement expects the emoji font size will have a different font size compared to the normal text, whether the text only contains emoji or also contains some text. 

I know this issue does not ask for that, but if we do the replacement at `ExpensiMark`, later on when we work on that improvement, we need somehow to tackle the case for the emoji only text and it is not feasible to do it at `ExpensiMark`. If we do the replacement in our App code, we can just easily change the tag to another tag that indicates the text only contains emoji.

</details>

```diff
diff --git a/src/pages/home/report/ReportActionItemFragment.js b/src/pages/home/report/ReportActionItemFragment.js
index c4f593517..4f34733cc 100644
--- a/src/pages/home/report/ReportActionItemFragment.js
+++ b/src/pages/home/report/ReportActionItemFragment.js
@@ -108,7 +108,8 @@ const ReportActionItemFragment = (props) => {
             // Only render HTML if we have html in the fragment
             if (!differByLineBreaksOnly) {
                 const editedTag = props.fragment.isEdited ? '<edited></edited>' : '';
-                const htmlContent = html + editedTag;
+                const htmlContent = EmojiUtils.encloseEmojisWithTag(html, 'emoji') + editedTag;
+                
                 return (
                     <RenderHTML
                         html={props.source === 'email'
diff --git a/src/libs/EmojiUtils.js b/src/libs/EmojiUtils.js
index d6ed8875e..dce4ad93c 100644
--- a/src/libs/EmojiUtils.js
+++ b/src/libs/EmojiUtils.js
@@ -245,6 +245,10 @@ function suggestEmojis(text, limit = 5) {
     return [];
 }
 
+function encloseEmojisWithTag(text, tag) {
+    return Str.replaceAll(text, CONST.REGEX.EMOJIS, g => `<${tag}>${g}</${tag}>`);
+}
+
 export {
     getDynamicHeaderIndices,
     mergeEmojisWithFrequentlyUsedEmojis,
@@ -253,4 +257,5 @@ export {
     replaceEmojis,
     suggestEmojis,
     trimEmojiUnicode,
+    encloseEmojisWithTag,
 };
```

Now, apply the style to the custom tag

```diff
diff --git a/src/styles/styles.js b/src/styles/styles.js
index 2b2a3498d..8e7d163af 100644
--- a/src/styles/styles.js
+++ b/src/styles/styles.js
@@ -1065,6 +1065,11 @@ const styles = {
         textDecorationLine: 'none',
     },
 
+    emojisText: {
+        fontStyle: 'normal', 
+        fontWeight: 'normal',
+    },
+
     onlyEmojisText: {
         fontSize: variables.fontSizeOnlyEmojis,
         lineHeight: variables.fontSizeOnlyEmojisHeight,
diff --git a/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js b/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
index 456cd1ecf..fcfe6b809 100755
--- a/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
+++ b/src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js
@@ -39,6 +39,10 @@ const customHTMLElementModels = {
         tagName: 'email-comment',
         mixedUAStyles: {whiteSpace: 'normal'},
     }),
+    'emoji': defaultHTMLElementModels.span.extend({
+        tagName: 'emoji',
+        mixedUAStyles: styles.emojisText,
+    }),
 };
 
 const defaultViewProps = {style: [styles.alignItemsStart, styles.userSelectText]};

```

The last problem is when we type emojis inside code (` `) tag, the text does not show, for example `😄`. The text will be converted to `<code><emoji>😄</emoji></code>`. We have a custom rendered for code tag and it's only taking a text inside the code tag, but the emoji is enclosed with `emoji` tag. The solution is to get the text inside of it.

The getEmojiTextInsideCode will iterate over the children inside the code tag and get the data (text) of it. For example, if we have this text `a😄`, the children will be `<TText data="a" />` and `<TText data="😄" />`.

```diff
diff --git a/src/components/HTMLEngineProvider/htmlEngineUtils.js b/src/components/HTMLEngineProvider/htmlEngineUtils.js
index 71a5e9267..fb588ff08 100644
--- a/src/components/HTMLEngineProvider/htmlEngineUtils.js
+++ b/src/components/HTMLEngineProvider/htmlEngineUtils.js
@@ -44,8 +44,13 @@ function isInsideComment(tnode) {
     return false;
 }
 
+function getEmojiTextInsideCode(tnode) {
+    return tnode.children.reduce((prev, curr) => prev + curr.data, '');
+}
+
 export {
     computeEmbeddedMaxWidth,
     isInsideComment,
     isCommentTag,
+    getEmojiTextInsideCode,
 };
diff --git a/src/components/InlineCodeBlock/index.js b/src/components/InlineCodeBlock/index.js
index af33b4468..222b7d877 100644
--- a/src/components/InlineCodeBlock/index.js
+++ b/src/components/InlineCodeBlock/index.js
@@ -1,6 +1,7 @@
 import React from 'react';
 import inlineCodeBlockPropTypes from './inlineCodeBlockPropTypes';
 import Text from '../Text';
+import * as HTMLEngineUtils from '../HTMLEngineProvider/htmlEngineUtils';
 
 const InlineCodeBlock = (props) => {
     const TDefaultRenderer = props.TDefaultRenderer;
@@ -12,7 +13,9 @@ const InlineCodeBlock = (props) => {
             <Text
                 style={{...props.boxModelStyle, ...props.textStyle}}
             >
-                {props.defaultRendererProps.tnode.data}
+                {props.defaultRendererProps.tnode.data
+                    ? props.defaultRendererProps.tnode.data
+                    : HTMLEngineUtils.getEmojiTextInsideCode(props.defaultRendererProps.tnode)}
             </Text>
         </TDefaultRenderer>
     );
diff --git a/src/components/InlineCodeBlock/index.native.js b/src/components/InlineCodeBlock/index.native.js
index 8431ce61e..04adad238 100644
--- a/src/components/InlineCodeBlock/index.native.js
+++ b/src/components/InlineCodeBlock/index.native.js
@@ -2,6 +2,7 @@ import React from 'react';
 import styles from '../../styles/styles';
 import WrappedText from './WrappedText';
 import inlineCodeBlockPropTypes from './inlineCodeBlockPropTypes';
+import * as HTMLEngineUtils from '../HTMLEngineProvider/htmlEngineUtils';
 
 const InlineCodeBlock = (props) => {
     const TDefaultRenderer = props.TDefaultRenderer;
@@ -17,7 +18,9 @@ const InlineCodeBlock = (props) => {
                     styles.codeWordStyle,
                 ]}
             >
-                {props.defaultRendererProps.tnode.data}
+                {props.defaultRendererProps.tnode.data
+                    ? props.defaultRendererProps.tnode.data
+                    : HTMLEngineUtils.getEmojiTextInsideCode(props.defaultRendererProps.tnode)}
             </WrappedText>
         </TDefaultRenderer>
     );
```

Finally, I thought [this](https://github.com/Expensify/App/issues/4114) improvement should be held waiting for this issue.

**Comment by tienifr** [comment]
thanks @bernhardoj for the additional insights 💪

I agree there’ll be some fixes here and there to make sure it works 100% in all cases, some of which were [mentioned earlier](https://github.com/Expensify/App/issues/14676#issuecomment-1413526293) and some you mentioned above.

Once the direction is confirmed we can dive deeper and resolve the remaining (potential) issues.

**Comment by tienifr** [comment]
@bernhardoj in the case of the 👨‍👨‍👦 emoji, it's indeed split into sub-emojis but it still renders correctly on the screen (since the sub-emojis are in the consecutive span tags and still render the correct composite emoji).

Just curious if you find any case of an emoji that has user-visible problem if it's not split by your updated regex?

**Comment by bernhardoj** [comment]
Surprisingly it worked on web, but that's not the case for native platform.

**Comment by MelvinBot** [comment]
@flodnv @anmurali @aimane-chnaif this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
@flodnv @anmurali @aimane-chnaif this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $2000

**Comment by MelvinBot** [comment]
@flodnv, @anmurali, @aimane-chnaif Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by flodnv** [comment]
We're still discussing with @aimane-chnaif and @rushatgabhane

**Comment by anmurali** [comment]
No update over the weekend.

**Comment by MelvinBot** [comment]
@flodnv @anmurali @aimane-chnaif this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by alexxxwork** [comment]
> We're still looking for best solution.

I don't have access to internal discussion. If you are still waiting for alternative solutions, could you please explain the point of discussion briefly? It seems that @bernhardoj 's soultion fits - it solves this problem and provides solution for another issue. Or it's just doesn't work on native?

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Emoji has italic style if put in `_` `_` in markdown.

### What is the root cause of that problem?
The emoji inside `<em>` tag has italic style.

### What changes do you think we should make in order to solve the problem?

As mentioned by @aimane-chnaif above, we can update the italic rule to exclude the emoji from the `em` tag. We can do that by adding a `post` function in the current italic rule:

```
post: (text) => {
                    return text.replace(/(([\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3)+)/g, '<em/>$1<em>')
                }
```

This will replace all emoji group inside the `<em>`, `</em>` by ` '<em/>$1<em>'`, which will achieve the desired behavior.

### What alternative solutions did you explore? (Optional)
NA

**Comment by tienifr** [comment]
> @tienifr I need this solution as well.

@aimane-chnaif sure.

This rule will work for all the cases you mentioned:

```
{
                /**
                 * Use \b in this case because it will match on words, letters,
                 * and _: https://www.rexegg.com/regex-boundaries.html#wordboundary
                 * The !_blank is to prevent the `target="_blank">` section of the
                 * link replacement from being captured Additionally, something like
                 * `\b\_([^<>]*?)\_\b` doesn't work because it won't replace
                 * `_https://www.test.com_`
                 * Use [\s\S]* instead of .* to match newline
                 */
                name: 'italic',
                regex: /(?!_blank")\b_((?!\s)[\s\S]*?\S)_\b(?![^<]*(<\/pre>|<\/code>|<\/a>|_blank))/g,
                replacement: (match, g1) => {
                    const replacedText =  `<em>${g1}</em>`;
                    const textWithEmojiExcluded = replacedText.replace(/(([\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3)+)/gu, '</em>$1<em>');
                    const textWithEmptyTagRemoved = textWithEmojiExcluded.replace(/<em><\/em>/g, '');
                    return textWithEmptyTagRemoved;
                }
            },
```
(pls note the change in `replacement` to handle the replacement steps)

**Comment by bernhardoj** [comment]
If we do [this](https://github.com/Expensify/App/issues/14676#issuecomment-1441443520), can we convert it back to the markdown 100% the same because now we have multiple `_`?

For example, we convert from this markdown `_a🥳b🥳c_` to the html `<em>a</em>🥳<em>b</em>🥳<em>c</em>`. When we convert it back to the html, we will get `_a_🥳_b_🥳_c_`.

**Comment by tienifr** [comment]
@aimane-chnaif in this case we will need to make the htmlToMarkdown conversion consistent with 1 style of putting the emoji. I think it should be converted to ‘ _hello✋world_’ in both cases. Both renders the same way for the user, so the redundant _ _ wrapping around the emoji is safe to be removed

**Comment by tienifr** [comment]
@aimane-chnaif it might not be possible since we're not preserving the order of the markdown in the html. Can you discuss in the Slack thread about this issue to see if we change our mind on which way we want to go? (if we use a custom emoji tag like Github does then we'll not have this problem)

> it might not be possible since we're not preserving the order of the markdown in the html

To workaround this I can imagine we can put some special properties in the `em` tag to indicate this, however it's like a workaround and can be buggy.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We don't want to apply bold and italic style to emoji.

### What is the root cause of that problem?
Emoji inside `_` and `*` markdown will receive bold and italic text style.

### What changes do you think we should make in order to solve the problem?


1. Change emoji regex to `/[\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3/gu`
2. Modify the text here by wrapping the emoji with `</em|b>emoji<em|b>` and remove empty tag.
https://github.com/Expensify/App/blob/835a9f1755f33c12ff425c7f06eece581364be49/src/pages/home/report/ReportActionItemFragment.js#L111

We can create a new util function, let's call it `unwrapEmojis`.

First, we need a new regex constructed from CONST.REGEX.EMOJIS to match all emojis inside bold or italic tag.

```
RegExp(`<(b|strong|em|i)>.*(${CONST.REGEX.EMOJIS.source})+.*<\\/\\1>`, CONST.REGEX.EMOJIS.flags)
```

Next, we match the whole text with the new regex and then replace the emojis with the tag mentioned above (step 2).

Last, we clean the empty tag.

It will look like this:

```
function unwrapEmojis (text):
    emojiRegex = RegExp(`<(b|strong|em|i)>.*(${CONST.REGEX.EMOJIS.source})+.*<\\/\\1>`, CONST.REGEX.EMOJIS.flags)
    return text.replace(emojiRegex, (match, g1):
        return match.replace(CONST.REGEX.EMOJIS, (match) => `</${g1}>${match}<${g1}>`).replace(new RegExp(`<${g1}></${g1}>`, 'g'), '')
    )
```

**Comment by MelvinBot** [comment]
@flodnv @anmurali @aimane-chnaif this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @aimane-chnaif is eligible for the Internal assigner, not assigning anyone new.

**Comment by tienifr** [comment]
> @tienifr we already discussed. that's why we suggest to fix this in render level

Sure, I misunderstood what you mean by `render level`. In this case we can just apply the sequence of regex mentioned previously in [here](https://github.com/Expensify/App/issues/14676#issuecomment-1446561832) to the rendering.

## Updated Proposal

### Please re-state the problem that we are trying to solve in this issue.
Emoji has italic style if put in `_` `_` in markdown.

### What is the root cause of that problem?
The emoji inside `<em>`/`<i>` tag in html has italic style. Same for bold.

### What changes do you think we should make in order to solve the problem?

As mentioned by @aimane-chnaif above, when rendering the html, we can exclude the emoji from the `em` and similar tags. We can do that adding the following function to process the html:
```
const excludeEmojisFromItalicAndBoldTags = (text) => {
    return text.replace(/<(em|i|b|strong)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi, (match, g1) => {
        const textWithEmojiExcluded = match.replace(/(([\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3)+)/gu, `</${g1}>$1<${g1}>`);
        const textWithEmptyTagRemoved = textWithEmojiExcluded.replace(new RegExp(`<${g1}></${g1}>`, 'g'), '');
        return textWithEmptyTagRemoved;
    });
}
```
this will be called with `htmlContent` as the param right before this line here https://github.com/Expensify/App/blob/79e8dd706f5760f398eaa1b94e2e2be97805fd6a/src/pages/home/report/ReportActionItemFragment.js#L112

`const htmlContent = excludeEmojisFromItalicAndBoldTags(html + editedTag);`

This will replace all emoji group inside the `<em>`, `</em>` by ` '</em>$1<em>'`, which will achieve the desired behavior.

We further refine this by removing redundant `<em></em>` that doesn't have any content inside. This can happen if an emoji happens to be right next to the leading and trailing `<em>`.

### What alternative solutions did you explore? (Optional)
Another way is instead of the regex above to handle the emojis, we can have a regex to replace all `non-emoji character group` with `<em>$1</em>` (the `<em>` or `</em>` will be omitted if the group is next to the start or end of the tag)

**Comment by bernhardoj** [comment]
There is an [issue](https://github.com/Expensify/App/issues/15424) that expects (only) emojis inside bold and italic should have larger size. But based on this [confirmation](https://github.com/Expensify/App/issues/4114#issuecomment-1448461796), it shouldn't. But (again), in this issue, we are trying to take out the emoji from the italic/bold tag, which technically makes the emojis not inside bold/italic anymore. So, I have updated my [proposal](https://github.com/Expensify/App/issues/14676#issuecomment-1446831092) with a very little bit changes in case we want to show larger emojis inside bold/italic. If not, then I think we can close this [issue](https://github.com/Expensify/App/issues/15424).

**Comment by anmurali** [comment]
Proposals still being evaluated

**Comment by flodnv** [comment]
I asked for more voices on one of our internal Slack channels

**Comment by flodnv** [comment]
Let's HOLD on that PR to see if it fixes it.

**Comment by flodnv** [comment]
Still HOLDing, that PR is in progress.

**Comment by flodnv** [comment]
That PR is now on HOLD with a bunch of conflicts 😂 moving this to Monthly...

**Comment by priya-zha** [comment]
Hi @flodnv. Wanted to know if this issue [here](https://github.com/Expensify/App/issues/17553) is same or not?

**Comment by flodnv** [comment]
This is still waiting on https://github.com/facebook/hermes/pull/968, which had some progress in the past month.

**Comment by anmurali** [comment]
https://github.com/facebook/hermes/pull/968 is still WIP

**Comment by flodnv** [comment]
Wow, it was merged!! 😂 What now?! Do you know how these PRs end up in RN releases @mountiny ?

**Comment by mountiny** [comment]
I believe there should be new 0.73 release coming up soon (October hopefully) cc @fabioh8010 do you have some better ETA?

**Comment by fabioh8010** [comment]
Hi @mountiny , the target date for RN 0.73 is 23 October ([more info here](https://github.com/reactwg/react-native-releases/discussions/64)). [And according to Meta's team](https://github.com/facebook/hermes/pull/968#issuecomment-1633011552), this Hermes improvement is expected to be available on RN 0.73 as well.

But please have in mind that ETA will likely be 23 October + time to update E/App to latest RN version.

**Comment by mountiny** [comment]
Cool thanks!

**Comment by flodnv** [comment]
Still waiting for RN 0.73

**Comment by flodnv** [comment]
Same.

**Comment by fabioh8010** [comment]
We also can watch for this PR which is updating the app to 0.73 -> https://github.com/Expensify/App/pull/31558

**Comment by flodnv** [comment]
Seems like we're almost there 🙏

**Comment by anmurali** [comment]
https://github.com/Expensify/App/pull/31558 is deployed to production. @flodnv @fabioh8010 - what is the next step here?

**Comment by flodnv** [comment]
The problem still exists, so it needs to be fixed (or, we can close this issue as won't fix)

**Comment by flodnv** [comment]
Still not convinced this is a ""bug"". Look at GH:

<img width="925" alt="image" src="https://github.com/Expensify/App/assets/4164921/3f2f31be-ad53-43da-a3a0-8ab6e99de215">

Seen in https://github.com/Expensify/App/pull/35552

>I think this should be fixed for consistency. 

Consistency with what?

**Comment by flodnv** [comment]
Hm, I see the bug on Desktop and web (Chrome, MacOS). Are you saying it does not exist on iPhone?

**Comment by flodnv** [comment]
Thanks, I think all that's left to test is iPhone? And maybe Android mWeb and iPhone mWeb

**Comment by flodnv** [comment]
Alright then, I can kind of buy the consistency argument 👍

**Comment by anmurali** [comment]
Adding in BZ backup when I am OOO

**Comment by stitesExpensify** [comment]
Wrong button :ohnothing:

I'm going to take this over as the internal engineer. We should hold on this until [this](https://github.com/Expensify/App/pull/35838) is merged because we're adding a new `<emoji>` tag that will change the solution here. 

I also wanted to mention that we are adding a requirement here to add a unit test. We have gone back and forth on whether emojis can be for formatted or not a few times, and we want to lock down that emojis should never be formatted.

**Comment by mallenexpensify** [comment]
Off hold! @stitesExpensify , do you still want this to be a Monthly?

**Comment by stitesExpensify** [comment]
I'm not sure why @eh2077 was added, @aimane-chnaif is already the c+ here right?

**Comment by eh2077** [comment]
@stitesExpensify I'm not sure if this is because @aimane-chnaif is OOO as they mentioned [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1710957534578839)

**Comment by eh2077** [comment]
> Wrong button :ohnothing:
> 
> I'm going to take this over as the internal engineer. We should hold on this until [this](https://github.com/Expensify/App/pull/35838) is merged because we're adding a new `<emoji>` tag that will change the solution here.
> 
> I also wanted to mention that we are adding a requirement here to add a unit test. We have gone back and forth on whether emojis can be for formatted or not a few times, and we want to lock down that emojis should never be formatted.

@tienifr @bernhardoj Would you like to update your proposals based on this https://github.com/Expensify/App/issues/14676#issuecomment-1967604879 ?

**Comment by bernhardoj** [comment]
Updated my [proposal](https://github.com/Expensify/App/issues/14676#issuecomment-1446831092), but it's only for sent message and not for live markdown preview.

**Comment by eh2077** [comment]
Reviewing proposals

**Comment by robertKozik** [comment]
Hi! I've stumbled across this issue when I was trying to solve problem with parsing emojis inside code blocks and HTML tags ( as HTML attributes) - ideally they should not be wrapped in the html tag. Will this issue cover this case as well?

**Comment by eh2077** [comment]
@robertKozik Thanks for your comment. Do you mean emojis inside code block have same issue (be formatted)? Or can you elaborate the issue you're facing?

**Comment by robertKozik** [comment]
I think we should not parse (wrap in `<emoji/>` tag) any emoji within code and code fences as adding any styling to that tag would lead to being formatted.

For example string ` ```🫰🏼``` ` will be parsed as:
```
<pre data-code-raw="<emoji>🫰🏼</emoji>">
    <emoji>🫰🏼</emoji>
</pre>
```
So in this case if we apply any formatting for the `emoji` tag we will format the emoji inside the code block as well.

Do you think this is the right issue for such a case?

**Comment by eh2077** [comment]
@robertKozik Thanks for the explanation. We'll revisit your feedback when we get a final solution.

@bernhardoj  I'm not sure if we should unwrap emojis from multiple-line code block given `<pre>` is a block type element.

**Comment by bernhardoj** [comment]
> I'm not sure if we should unwrap emojis from multiple-line code block given <pre> is a block type element.

Agree, unwrapping emoji tag also wouldn't match the expectation of @robertKozik, because if I understand it correctly, we don't want to apply the `<emoji>` tag when it's inside a codefence.

The solution in this issue will render `*test😄test*` as `<b>test</b><emoji>😄</emoji><b>test</b>`

**Comment by eh2077** [comment]
>if I understand it correctly, we don't want to apply the `<emoji>` tag when it's inside a codefence.

@bernhardoj But, currently, we do wrap emojis inside `<code>` and `<pre>` with `<emoji>` tag. See below comparisons with slack

| Expensify  | Slack |
| ------------- | ------------- |
| <video width="30" src="https://github.com/Expensify/App/assets/117511920/9ca78b6b-a7b0-4d06-b952-879158bd2e74">  | <img width="400" alt="image" src="https://github.com/Expensify/App/assets/117511920/62ee7f54-9b8d-4329-9702-8957712956ae">|

It seems the tooltip of emoji inside `<code>` is not shown up. Maybe it's a bug?

@stitesExpensify I saw your https://github.com/Expensify/App/issues/34307#issuecomment-1981280845
> We don't want emojis to be affected by any markdown, but that is a separate issue technically https://github.com/Expensify/App/issues/14676

So, we just want emojis to be large size (when with emojis only) or normal size (when with text) inside any markdown syntax, right?


@bernhardoj I'm not sure if it's feasible to solve this issue by only adding css styles to `<emoji>` tag when rendering them.

**Comment by bernhardoj** [comment]
>  But, currently, we do wrap emojis inside <code> and <pre> with <emoji> tag.

Yes, I'm saying @robertKozik wants to change so emoji tag is not present inside the inline code or code fence which is out of scope of this issue.

> I'm not sure if it's feasible to solve this issue by only adding css styles to <emoji> tag when rendering them.

We can't do that to hide strikethrough.

**Comment by eh2077** [comment]
> 
> > I'm not sure if it's feasible to solve this issue by only adding css styles to  tag when rendering them.
> 
> We can't do that to hide strikethrough.

@stitesExpensify Do we want strikethrough on an emoji? I found that Slack doesn't do that 

<img width="113" alt="image" src="https://github.com/Expensify/App/assets/117511920/e19655ea-9b9d-47ac-ae1e-2f62054c7527">

**Comment by stitesExpensify** [comment]
> So, we just want emojis to be large size (when with emojis only) or normal size (when with text) no matter it's wrapped by any markdown syntax, right?

> Do we want strikethrough on an emoji? I found that Slack doesn't do that

We want the behavior to be exactly the same as slack

**Comment by bernhardoj** [comment]
My proposal already covers the strikethrough case.

**Comment by eh2077** [comment]
> > I'm not sure if it's feasible to solve this issue by only adding css styles to  tag when rendering them.
> 
> We can't do that to hide strikethrough.

I see. But I'd like to ask if you would like to research this method. If it works, I think it will provide benefit of not changing the markdown when editing the comment. While your current solution has a downside in the editing case.

**Comment by bernhardoj** [comment]
> While your current solution has a downside in the editing case.

The fix only applies to TextCommentFragment, so I don't see why it can affect editing.

> But I'd like to ask if you would like to research this method

I did and it's not possible except there is something I miss. Even if you set line through to unset to the children, the parent line through will still be shown. We can apply display: inline block to the children, but it's only available on web.

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

 Do not allow emojis to be formatted (ie italics) 

### What is the root cause of that problem?

I feel this is feature request, emoji will follow the font style by default 

### What changes do you think we should make in order to solve the problem?

To change the emoji style in the submitted text, modify [EmojiWithTooltip in EmojiRenderer](https://github.com/Expensify/App/blob/f82f8636191ab8e70154e4ab7ed045f0452023d0/src/components/HTMLEngineProvider/HTMLRenderers/EmojiRenderer.tsx#L10-L13).

```
        <EmojiWithTooltip
            style={[style, styles.cursorDefault, styles.emojiStyle]}
            emojiCode={'data' in tnode ? tnode.data : ''}
        />
```

The `emojiStyle` will contain `fontWeight: 'normal'` and `fontStyle: 'normal'`

For live markdown preview, we can allow the `emoji` tag to have the `fontWeight` and `fontStyle` properties. These properties will override the formatting from the italic/bold wrapper.

Add to `src/hooks/useMarkdownStyle.ts`:

```
emoji: {
                fontStyle: 'normal',
                fontWeight: 'normal',
                fontSize: 12,
            },
```

There are some changes required in the live markdown to properly handle emoji unformatting, including:
1. Make **bold** and *italic* type have higher priority than `emoji` type. This will ensure this test case passed: `_emoji_`
Change [getTagPriority](https://github.com/Expensify/react-native-live-markdown/blob/c7595572ff4f2a4b8c95a6110a05de62b2d8dbe4/parser/index.ts#L171-L180) to:
```
function getTagPriority(tag: string) {
  switch (tag) {
    case 'blockquote':
      return 2;
    case 'h1':
      return 1;
    case 'italic':
      return 1;
    case 'bold':
      return 1;
    default:
      return 0;
  }
```

2. For android, ranges must be readjusted to avoid overlapping with the emoji range, as overlapping the same styleSpan will maintain the formatting of our emoji. Modify [this code](https://github.com/Expensify/react-native-live-markdown/blob/c7595572ff4f2a4b8c95a6110a05de62b2d8dbe4/android/src/main/java/com/expensify/livemarkdown/MarkdownUtils.java#L92-L96) to:

```
....
      JSONArray ranges = new JSONArray(output);
      JSONArray adjustedRanges =  adjustRanges(ranges);

      for (int i = 0; i < adjustedRanges.length(); i++) {
            JSONObject range = adjustedRanges.getJSONObject(i);
            String type = range.getString("type");
            int start = range.getInt("start");
    ......
```

<details>
<summary>
This is the function we can use to handle overlapping style (adjustRanges)
</summary>

```java
  public JSONArray adjustRanges(JSONArray ranges) throws JSONException {
    String emojiFontWeight = mMarkdownStyle.getEmojiFontWeight();
    String emojiFontStyle = mMarkdownStyle.getEmojiFontStyle();

    if ((emojiFontWeight == null && emojiFontStyle == null) || (emojiFontWeight.equals("bold") && emojiFontStyle.equals("italic"))) {
      return ranges;
    }

    JSONArray newRanges = new JSONArray();
    for (int i = 0; i < ranges.length(); i++) {
      JSONObject range = ranges.getJSONObject(i);
      String type = range.getString("type");
      int start = range.getInt("start");
      int length = range.getInt("length");
      int end = start + length;

      if (type.equals("emoji")) {
        // If the type is emoji, remove any overlapping bold or italic range
        for (int j = 0; j < newRanges.length(); j++) {
          JSONObject existingRange = newRanges.getJSONObject(j);
          String existingType = existingRange.getString("type");
          int existingStart = existingRange.getInt("start");
          int existingLength = existingRange.getInt("length");
          int existingEnd = existingStart + existingLength;

          boolean isBold = "bold".equals(existingType) && emojiFontWeight != null && !"bold".equals(emojiFontWeight);
          boolean isItalic = "italic".equals(existingType) && emojiFontStyle != null && !"italic".equals(emojiFontStyle);

          if ((isBold || isItalic)
              && (existingStart < end && existingEnd > start)) {
            // If the emoji range is completely within the existing range, split the existing range
            if (start > existingStart && end < existingEnd) {
              existingRange.put("length", start - existingStart);
              JSONObject newRange = new JSONObject();
              newRange.put("type", existingType);
              newRange.put("start", end);
              newRange.put("length", existingEnd - end);
              newRanges.put(newRange);
            } else {
              // Otherwise, adjust the existing range to not overlap with the emoji range
              if (start > existingStart) {
                existingRange.put("length", start - existingStart);
              } else if (end < existingEnd) {
                existingRange.put("start", end);
                existingRange.put("length", existingEnd - end);
              } else {
                newRanges.remove(j);
                j--;
              }
            }
          }
        }
      }
      newRanges.put(range);
    }
    return newRanges;
  }
```

</details>
We should also allow overriding fontWeight and fontStyle for iOS code and manage overlapping, if necessary.

### What alternative solutions did you explore? (Optional)

**Comment by eh2077** [comment]
@bernhardoj

> > While your current solution has a downside in the editing case.
> 
> The fix only applies to TextCommentFragment, so I don't see why it can affect editing.

Oh, yes. I overlooked it. This method works well when editing and `Copy to clipboard` for pasting into the composer. But there's a drawback when copying comment via `Ctrl/Cmd + C`. See clips below

| unwrapEmojis  | Production |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/117511920/e605b6fb-94b3-4a91-9d8d-fd5f88686cb0">  | <video src="https://github.com/Expensify/App/assets/117511920/b68e0e8a-6baa-44a8-8ef1-242004ff5e65">|










> > But I'd like to ask if you would like to research this method
> 
> I did and it's not possible except there is something I miss. Even if you set line through to unset to the children, the parent line through will still be shown. We can apply display: inline block to the children, but it's only available on web.

Thanks for sharing.

**Comment by eh2077** [comment]
@wildan-m Thanks for your proposal! I'll need taking sometime to look into it.

**Comment by bernhardoj** [comment]
@eh2077 I see. We need a different approach then.

**Comment by wildan-m** [comment]
Proposal [Updated](https://github.com/Expensify/App/issues/14676#issuecomment-2037745820)

Add fix for emoji inside `code` block.

**Comment by wildan-m** [comment]
Proposal [Updated](https://github.com/Expensify/App/issues/14676#issuecomment-2037745820)

`inlineCodeBlock`: Optimize loop, handle containOnlyEmojis case

**Comment by eh2077** [comment]
Reviewing the proposal from @wildan-m

**Comment by eh2077** [comment]
@wildan-m Will your solution be simpler if we only need to handle emojis inside bold/italics/strikethrough markdown? Like
```
*bold 😄 bold*
_italic 😄 italic_
~strikethrough 😄 strikethrough~
```

| Expensify  | Slack |
| ------------- | ------------- |
| <img width="215" alt="image" src="https://github.com/Expensify/App/assets/117511920/5f38b818-cbb9-4ba2-b57e-1234986b2bbe">  | <img width="239" alt="image" src="https://github.com/Expensify/App/assets/117511920/00c3bdcf-0ef6-4c7b-9aa1-59234759346e"> |

**Comment by eh2077** [comment]
@wildan-m I tried to follow your solution. 
> To change the emoji style in the submitted text, modify [EmojiWithTooltip in EmojiRenderer](https://github.com/Expensify/App/blob/f82f8636191ab8e70154e4ab7ed045f0452023d0/src/components/HTMLEngineProvider/HTMLRenderers/EmojiRenderer.tsx#L10-L13).

Can this solve emojis in bold/italics/strikethrough markdown as described https://github.com/Expensify/App/issues/14676#issuecomment-2047490260 ? Or can you elaborate how to override styles for emojis inside bold/italics/strikethrough markdown? 

I think we don't need to consider live markdown at this moment because if IRCC, that project is maintain by SWM team.

**Comment by eh2077** [comment]
> > So, we just want emojis to be large size (when with emojis only) or normal size (when with text) no matter it's wrapped by any markdown syntax, right?
> 
> > Do we want strikethrough on an emoji? I found that Slack doesn't do that
> 
> We want the behavior to be exactly the same as slack

@stitesExpensify Just want to be clear, should we also display emoji code in code or code block? Just like Slack

| Expensify  | Slack |
| ------------- | ------------- |
| <video width="30" src="https://github.com/Expensify/App/assets/117511920/9ca78b6b-a7b0-4d06-b952-879158bd2e74">  | <img width="400" alt="image" src="https://github.com/Expensify/App/assets/117511920/62ee7f54-9b8d-4329-9702-8957712956ae">|

There're comments related to this discussion from another issue, see 
- https://github.com/Expensify/App/issues/34307#issuecomment-1980739109
- https://github.com/Expensify/App/issues/34307#issuecomment-2033020632
- https://github.com/Expensify/App/issues/34307#issuecomment-2033198292

I think displaying emoji code text inside code/code block makes more sense. 

cc @robertKozik

**Comment by wildan-m** [comment]
@eh2077 Proposal [Updated](https://github.com/Expensify/App/issues/14676#issuecomment-2037745820)

Exclude strikethrough formatting to the emoji submitted text.

Result for:
```
*bold 😄 bold*
_italic 😄 italic_
~strikethrough 😄 strikethrough~
```

<img width="455" alt="image" src="https://github.com/Expensify/App/assets/11847279/6253ee21-cb5b-4f01-8c16-826c2bc600cd">

**Comment by wildan-m** [comment]
@eh2077 works pretty well in my test

<img width="302" alt="image" src="https://github.com/Expensify/App/assets/11847279/1db9d3eb-06e7-4374-9bd3-829433b47b29">

Can you try to update the chrome browser?

If you need to test my branch you can check it [here](https://github.com/Expensify/App/compare/main...wildan-m:App:wildan/fix-14676-prevent-format-emoji).

**Comment by eh2077** [comment]
@wildan-m I applied these changes but still not working on mobile Chrome

<img width="988" alt="image" src="https://github.com/Expensify/App/assets/117511920/f2dafe36-e968-4ab7-ace2-85e15fade01f">

> Can you try to update the chrome browser?

If the older version Chrome is an issue, can you specify the compatibility notes of the style you're using?

**Comment by wildan-m** [comment]
@eh2077 Proposal [Updated](https://github.com/Expensify/App/issues/14676#issuecomment-2037745820)

Add compatibility notes:
> [!Important]
> The `display: inline-flex` will work on mobile Chrome version [123 or the latest](https://caniuse.com/?search=inline-flex) If you use an Android emulator without Google Play, there is a chance that you are using a version below 123. Try using an emulator with Google Play for a smoother Chrome update process.

**Comment by eh2077** [comment]
@wildan-m Thanks for the update. I'll do more tests and feedback later

**Comment by wildan-m** [comment]
@eh2077 not sure if it's affecting the result but your https is being ignored. Maybe you can try using https://127.0.0.1:8082 or running the emulator from `npm run startAndroidEmulator`?

<img width="100" alt="image" src="https://github.com/Expensify/App/assets/11847279/ed9806de-c6da-4db0-8f64-41d1fb31d11f">

**Comment by eh2077** [comment]
@wildan-m I got same result after setting up https
<img width="346" alt="image" src="https://github.com/Expensify/App/assets/117511920/d94914fc-d09b-4767-a33a-ae47cda04f9b">

I'll test a new emulator with `Google Play`.

**Comment by eh2077** [comment]
### Review summary

---

@bernhardoj 's [proposal](https://github.com/Expensify/App/issues/14676#issuecomment-1446831092) to unwrap emojis has drawback when copy and paste comment via `Ctrl/Cmd + C`, see https://github.com/Expensify/App/issues/14676#issuecomment-2038743930

---

I verified that @wildan-m 's [proposal](https://github.com/Expensify/App/issues/14676#issuecomment-2037745820) to add style `styles.emojiDefault`
```css
        emojiDefault: {
            fontStyle: 'normal',
            fontWeight: 'normal',
            ...display.dInlineFlex,
            textDecorationLine:'none'
        },
```
works on all platforms when testing following markdown
```markdown
*bold 😄 bold*
_italic 😄 italic_
~strikethrough 😄 strikethrough~
```
<details>
<summary>Click to check demo for Web</summary>

<img width="388" alt="web" src="https://github.com/Expensify/App/assets/117511920/f1758f76-1661-4fc7-9556-9a7cb3e515c4">

</details>

<details>
<summary>Click to check demo for Desktop</summary>

<img width="1312" alt="desktop" src="https://github.com/Expensify/App/assets/117511920/0f431520-e326-4887-8389-899fc2b901df">


</details>

<details>
<summary>Click to check demo for Android (App and mobile Chrome) </summary>

<img width="321" alt="mobile-chrome" src="https://github.com/Expensify/App/assets/117511920/2fa90d67-7524-4707-8ba4-a28c2ae53b2a">

<img width="349" alt="image" src="https://github.com/Expensify/App/assets/117511920/4814d72f-019e-4772-b871-84c864f04efe">

</details>


<details>
<summary>Click to check demo for iOS (App and mobile Safari) </summary>
<img width="321" alt="iOS" src="https://github.com/Expensify/App/assets/117511920/2870dd9f-47af-44fc-b229-7f5814892831">
<img width="321" alt="mobile-safari" src="https://github.com/Expensify/App/assets/117511920/27b352ad-2b14-4041-af85-ed34c689870c">


</details>

> [!IMPORTANT]  
> On mobile Chrome, the style `display: inline-flex` only works on [123 or the latest](https://caniuse.com/?search=inline-flex)

> [!NOTE]  
> Live markdown in composer should be handled by [react-native-live-markdown](https://github.com/Expensify/react-native-live-markdown) project.

---

@stitesExpensify If you also agree with https://github.com/Expensify/App/issues/14676#issuecomment-2051355327 to display emoji code text inside code/code block markdown like Slack, then I think we can go with @wildan-m 's [proposal](https://github.com/Expensify/App/issues/14676#issuecomment-2037745820) to move this issue forward. We can handle the new feature to display emoji code text separately.


🎀👀🎀 C+ reviewed

**Comment by wildan-m** [comment]
@eh2077 The [PR is ready for review](https://github.com/Expensify/App/pull/40617)

**Comment by eh2077** [comment]
@robertKozik The app PR https://github.com/Expensify/App/pull/40617 has been shipped to production. Can you help with the live markdown part? Or any suggestions for asking people to help with it? Thank you.

**Comment by robertKozik** [comment]
Hi @eh2077 ! To double check the expected behaviour is to not format emojis at all? If we could get the github issue for it we will handle it within Live markdown team

**Comment by eh2077** [comment]
@robertKozik 

> To double check the expected behaviour is to not format emojis at all? 

Yes, that's right.

> If we could get the github issue for it we will handle it within Live markdown team

Do you mean a GH issue of the App repo? If so, then I think you can consider just use this one. Or maybe @stitesExpensify can help create a new issue.

**Comment by robertKozik** [comment]
Yeah E/App GH one, but I'm OK with either creating one or sticking to this one 🤷🏼‍♂️ whatever feels prefferable

**Comment by wildan-m** [comment]
@stitesExpensify, considering that the unformatted emoji from submitted text part has been successfully resolved and has completed its regression period, would it be possible to separate this issue from any other pending tasks? Decoupling these tasks might help us avoid delays.

**Comment by eh2077** [comment]
### Checklist

- [x] [@eh2077] The PR that introduced the bug has been identified. Link to the PR: _There's no a specific PR causes this issue as it exists when we support these formattings._
- [x] [@eh2077] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- [x] [@eh2077] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- [x] [@eh2077] Determine if we should create a regression test for this bug. **_Yes_**
- [x] [@eh2077] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

### Regression Test Proposal
- Go to any chat
- Paste the following markdown.

Markdown 1
```
*bold 😄 bold*
_italic 😄 italic_
~strikethrough 😄 strikethrough~
```

Markdown 2

```
`text🟥`
 
*`bold🟥`*

_`italic🟥`_

~`strikethrough🟥`~
```
- Verify that all of the emojis not affected by the formatting

- Do we agree 👍 or 👎

**Comment by mallenexpensify** [comment]
Contributor: @wildan-m paid $500 via Upwork
Contributor+: @eh2077 paid $500 via Upwork.  

Thanks for the checklist @eh2077 TestRail GH created
- https://github.com/Expensify/Expensify/issues/396229

@priya-zha , can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01cb7cfe875420b74e

**Comment by priya-zha** [comment]
@mallenexpensify Accepted. Thanks

**Comment by robertKozik** [comment]
@eh2077 @stitesExpensify What's the decision about `react-native-live-markdown` part of this issue? Should I link my PR with this issue or new will be created?

**Comment by mallenexpensify** [comment]
@priya-zha paid, will leave open til @robertKozik get's his question answered.

**Comment by robertKozik** [comment]
bumping

**Comment by mallenexpensify** [comment]
@aimane-chnaif , I'd recommend you fully follow the [process laid out in the C+ doc. ](https://docs.google.com/document/d/1BvohU05MTaHnjOD_vwJv_aqDAirv-ChkyRnKCAvOVyQ/edit#bookmark=id.25bt9k50o02o)

**Comment by mallenexpensify** [comment]
From @robertKozik above. 
> @eh2077 @stitesExpensify What's the decision about react-native-live-markdown part of this issue? Should I link my PR with this issue or new will be created?

@aimane-chnaif 👀 above too, if you want to request compensation

**Comment by mallenexpensify** [comment]
@stitesExpensify 👀 please on he above.  
Also.. going to put the below issue on hold pending this, comment if you disagree. 
- https://github.com/Expensify/App/issues/43322#issuecomment-2155557559

**Comment by stitesExpensify** [comment]
Apologies for the delay I've been OOO for a while now 😅 

Let's just link the PR to here @robertKozik 

> Also.. going to put the below issue on hold pending this, comment if you disagree.

Nope makes sense to me

**Comment by stitesExpensify** [comment]
@robertKozik just wanted to make sure this is on your radar. Can you confirm that the solution is being worked on?

**Comment by BartoszGrajdek** [comment]
Hi, I’m Bartosz from SWM `react-native-live-markdown` team. Robert is currently working on some other things, so I'll be taking this one over. I believe the fix for this bug is ready, we just have a few other issues with higher priority, so we'll update you on that 🫡

**Comment by stitesExpensify** [comment]
Great, thanks!

**Comment by mallenexpensify** [comment]
Assigned to @BartoszGrajdek , bumped to weekly and added to Live Markdown project. Thanks @BartoszGrajdek !

**Comment by BartoszGrajdek** [comment]
I still haven't managed to find time to work on this yet - I didn't forget about it though! Too busy managing all live markdown related issues ATM, sorry! 😅

**Comment by mallenexpensify** [comment]
Thanks for posting @BartoszGrajdek, keep focusing on what ya need to, this isn't the highest priority issue, just one we want to get wrapped up when we can since it's been open for so long.

**Comment by stitesExpensify** [comment]
No update, not a priority

**Comment by stitesExpensify** [comment]
@BartoszGrajdek where are we at with prioritization these days?

**Comment by BartoszGrajdek** [comment]
> BartoszGrajdek where are we at with prioritization these days?

We have much more time to handle things now, sorry for the wait 🙌 

@Skalakid decided to investigate if there's a cleaner way to handle this problem. If that won't be the case we'll follow through with the PR created by Robert after making some adjustments.

**Comment by stitesExpensify** [comment]
@Skalakid could you please provide an update?

**Comment by Skalakid** [comment]
@stitesExpensify sorry, I had to prioritize Live Markdown refactor PR and inline images feature. However, I prepared a small fix for the web. We can just add some styles to emoji default styles that will overwrite italic and other ones, like `fontStyle: normal`. Now we have to investigate this issue on the Android native side, since iOS appear to work properly now

**Comment by stitesExpensify** [comment]
Hey @Skalakid any more updates? :)

**Comment by stitesExpensify** [comment]
Bump :)

**Comment by Skalakid** [comment]
Hello, unfortunately we don't have any capacity to work on this issue now, because we are focused on fixing issues with higher priority

**Comment by mallenexpensify** [comment]
Thanks @Skalakid , I agree this isn't the highest priority.  I was going to _do something_ then realized we have 300 comments (which I didn't read all of before posting this) on this issue so I wanted to pause and reevaluate.  Should we
1. Bump this to monthly and the SWM can maybe get to it at some point 
2. Add `Help Wanted` to see if a contributor will want to work on
3. Add `Design` label to get feedback from them (ie. if they don't _care_, we _could_ close this)
4. Close it for now and maybe readdress later. 

cc @Skalakid , @BartoszGrajdek , @stitesExpensify @eh2077 @wildan-m

**Comment by Skalakid** [comment]
@mallenexpensify I think we can go with option number 1, we are finishing our current work and might have some time to investigate this issue on the Android native side

**Comment by mallenexpensify** [comment]
Will keep checking back ~monthly.

**Comment by stitesExpensify** [comment]
Hey @Skalakid any chance you got around to it this month?

**Comment by Skalakid** [comment]
Hello, I will take a look at it 👀

**Comment by tomekzaw** [comment]
FYI we added `emoji` range in live-markdown parser in https://github.com/Expensify/react-native-live-markdown/pull/233

**Comment by Skalakid** [comment]
Hi, here is the [PR](https://github.com/Expensify/react-native-live-markdown/pull/534) with the fix 😄

**Comment by stitesExpensify** [comment]
nice, that's awesome!!

**Comment by stitesExpensify** [comment]
@Skalakid what is the next step here?

**Comment by Skalakid** [comment]
Hi, we are waiting for this [PR](https://github.com/Expensify/react-native-live-markdown/pull/553) to be merged and after it's done we will merg [my PR](https://github.com/Expensify/react-native-live-markdown/pull/534)

**Comment by stitesExpensify** [comment]
Hey @Skalakid looks like your PR was closed, and we are moving things to expensimark, is that right?

**Comment by Skalakid** [comment]
Oh sorry, I totally forgot to write update under this issue. We've opened another PR that moves splitting logic from Android native code to TypeScript - [new PR](https://github.com/Expensify/react-native-live-markdown/pull/591). It has been merged and we had to disable splitting logic for web in [another PR](https://github.com/Expensify/react-native-live-markdown/pull/598). Both of the are deployed and now we are bumping the live markdown version to include these changes in the E/App - https://github.com/Expensify/App/pull/55509

**Comment by stitesExpensify** [comment]
Hey @Skalakid just checking in again here 😄  Where are we at?

**Comment by Skalakid** [comment]
The issue has been fixed and [the PR with the bump](https://github.com/Expensify/App/pull/55509) has been merged

**Comment by stitesExpensify** [comment]
Sorry, I'm not following. So what are the next steps?

**Comment by Skalakid** [comment]
I think the issue is resolved and we can close this one :D

**Comment by mallenexpensify** [comment]
<img width="201" alt="Image" src="https://github.com/user-attachments/assets/ff32818a-fdc9-48b8-b751-f6f3ad3ffb85" />

eeeps, this is a doozy.  Can contributors help me figure out who is owed what? (espec if there are multiple PRs and payments)

Thx

**Comment by eh2077** [comment]
@mallenexpensify I think you can just close this as we have been paid here https://github.com/Expensify/App/issues/14676#issuecomment-2111384153 and @Skalakid is from agency.

**Comment by mallenexpensify** [comment]
Thanks, woah... payment was from _forever_ ago 
Thanks @eh2077
